### DISCUSSION TITLE
How to get userId on api Endpoints while using next-auth

### DISCUSSION THREAD

**Discussion by youngnishant** [nextauthjs/next-auth#536] [Help] [answered]
*Please refer to the [documentation](https://next-auth.js.org/getting-started/introduction), the [example project](https://github.com/iaincollins/next-auth-example) and existing issues before creating a new issue.*

**Your question**
How to get userID on api Endpoints while using next-auth ?
**What are you trying to do**
A description of what you are trying to do.
While fetching data from mongodb atlas i want to use populate which requires UserID.

**Documentation feedback**
*Documentation refers to searching through [online documentation](https://next-auth.js.org), code comments and issue history. The example project refers to [next-auth-example](https://github.com/iaincollins/next-auth-example).*

* [x] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [x] Found the example project helpful
* [x] Could not find relevant documentation
* [ ] Did not find the example project helpful

**Comment by jcheese1** [comment]
```
const session = await getSession({ req });

session.id // userId
```

Is this it?

**Reply by ZendeAditya** [reply]
Not working

**Comment by nyedidikeke** [comment]
May be related to #535.

**Comment by ghost** [comment]
oh yeah. thats what I did as well, forgot that part.
tldr:

in [...nextauth].js:
```
callbacks: {
   session: async (session, user) => {
      session.id = user.id
      return Promise.resolve(session)
}
}
```

**Reply by halp1** [reply]
For modern nextauth, use this instead:
```javascript
callbacks: {
    session: async ({ session, user }) => {
      if (session?.user) {
        session.user.id = user.id;
      }
      return session;
    },
}
```

**Comment by youngnishant** [comment]
@yuyaoiwa-houzz  -- still getting error 

[next-auth][error][SESSION_ERROR] [
  TypeError: Cannot read property 'id' of undefined
      at Object.session (webpack-internal:///./pages/api/auth/[...nextauth].js:32:25)

**Comment by youngnishant** [comment]
@nyedidikeke I'm getting user as undefined

**Comment by ndom91** [comment]
@youngnishant can you please post the relevant code?

**Comment by youngnishant** [comment]
callbacks: {
   session: async (session, user) => {
      console.log(user)
      return Promise.resolve(session)
}
}

@ndom91

**Comment by jcheese1** [comment]
Maybe your DB isnt configured correctly?

**Comment by youngnishant** [comment]
@yuyao17 actually I'm using mongodb and checked it, it's working properly

**Comment by mcunningham** [comment]
@youngnishant are you using JWT? I had a similar issue using mongodb and followed the docs at [https://next-auth.js.org/configuration/callbacks](https://next-auth.js.org/configuration/callbacks) which states...

> If you want to pass data such as an Access Token or User ID to the browser when using JSON Web Tokens, you can persist the data in the token when the jwt callback is called, then pass the data through to the browser in the session callback.

I did so accordingly and was able to get a user id into my session object by using the below 2 callbacks:

1. I used the jwt callback to extract the user id (which is the mongodb _id from the users collection) and passed it to the token using `token.uid`

```javascript
jwt: async (token, user, account, profile, isNewUser) => {
      if (user) {
        token.uid = user.id;
      }
      return Promise.resolve(token);
    }
```

2. I then grabbed the token property I just set  and passed the data through to the browser in the session callback 
```javascript
 session: async (session, user) => {
      session.user.uid = user.uid;
      return Promise.resolve(session);
    }
```

I am then able to access `session.user.uid` when using the `useSession()` hook.

**Reply by asibilia** [reply]
So when I do the same thing I only get the `user` argument from the `jwt(token, user)` callback on the first time it's called (sign in). After that, every time `session()` is called `user` is `undefined` in `jwt(token, user)` which means it'll be `undefined` in the `session()` callback as well.

**Reply by abdullahtariq1171** [reply]
Hi @asibilia, I'm seeing same behaviour. Receiving `user` only for first call inside `jwt()` and nothing afterwards.  Wondering how did you solve it.

**Comment by youngnishant** [comment]
@shufflemattdev  actually I'm not using JWT ?

**Comment by youngnishant** [comment]
I'm glad that you tried to help 😊😄

**Comment by benoror** [comment]
Hey @youngnishant I am facing the exact same issue, can't fin the _id anywhere. Did you end up fixing it?

**Comment by benoror** [comment]
Note: If I disable `jwt` it does indeed include the `id` in the session (by injecting it in the callback)

**Comment by youngnishant** [comment]
@benoror I've implemented my own jwt authentication by taking proper security measures ✌️

**Comment by philippetedajo** [comment]
Try this:

// pages/api/auth/[...nextauth].js

callbacks: {
  session: async (session, user) => {
    session.userId = user.id;    
    return Promise.resolve(session);
  }
}
Then you can retrieve the userId from the session object using the getSession helper function:

const session = await getSession({ req });
const { userId } = session;

**Comment by goleary** [comment]
If you're using TypeScript then you'll also want to augment the default next-auth types by adding a file `types/next-auth.d.ts` that looks something like this:
```ts
import { DefaultUser } from "next-auth";

declare module "next-auth" {
  /**
   * Returned by `useSession`, `getSession` and received as a prop on the `SessionProvider` React Context
   */
  interface Session {
    user: DefaultUser & {
      id: string;
    };
  }
}
```

**Reply by noynek** [reply]
Much appreciated. I was getting lost on how to augment the TypeScript and this is exactly what I needed.

**Reply by ruohola** [reply]
Importing `NextAuth` was not needed actually.

**Reply by goleary** [reply]
@ruohola valid - removed that from the code block above.

Thanks!

**Reply by sayinmehmet47** [reply]
@goleary I added that file like you mentioned.
![image](https://user-images.githubusercontent.com/75525090/147247357-c77ed249-e796-4d50-9839-e5ab1add2b08.png)
I still can not take userId in data from useSession.
![image](https://user-images.githubusercontent.com/75525090/147247606-cd043658-8dc3-49cd-a7b2-d6e43a3f3e6e.png)
I am using typescript in nextauth

**Reply by ruohola** [reply]
@sayinmehmet47 See [my comment](https://github.com/nextauthjs/next-auth/discussions/536#discussioncomment-1863534), it should make this clear.

**Comment by ruohola** [comment, marked as answer]
I'm using `next-auth` `4.0.5`. Here's how I solved it (mostly combining advice from this thread):

**[EDIT] See my updated, simplified solution:** https://github.com/nextauthjs/next-auth/discussions/536#discussioncomment-1932922

`src/pages/api/auth/[...nextAuth].ts`:
```
export default NextAuth({
  ...
  callbacks: {
    session: async ({ session, token }) => {
      if (session?.user) {
        session.user.id = token.uid;
      }
      return session;
    },
    jwt: async ({ user, token }) => {
      if (user) {
        token.uid = user.id;
      }
      return token;
    },
  },
  session: {
    strategy: 'jwt',
  },
  ...
});
  ```

`types.d.ts` (file in project root):

```
import type { DefaultUser } from 'next-auth';

declare module 'next-auth' {
  interface Session {
    user?: DefaultUser & {
      id: string;
    };
  }
}

declare module 'next-auth/jwt/types' {
  interface JWT {
    uid: string;
  }
}
```

**Reply by sayinmehmet47** [reply]
![image](https://user-images.githubusercontent.com/75525090/147250610-b9d95479-3ca7-4c95-bf09-c0f0bfed3bde.png)
still not turns any id. 

Also I added that to [...nextAuth].ts
```js
 callbacks: {
      session: async ({ session, token }) => {
        if (session?.user) {
          session.user.id = token.uid;
        }
        return session;
      },
      jwt: async ({ user, token }) => {
        if (user) {
          token.uid = user.id;
        }
        return token;
      },
    },
```

**Reply by ruohola** [reply]
@sayinmehmet47 Did you sign in again to get a fresh JWT cookie?

**Reply by ruohola** [reply]
@sayinmehmet47 Try to step through the `session` and `jwt` callback functions with a debugger, or add some console.logs there to see what's happening.

**Reply by ruohola** [reply]
Actually I was able to simplify this even further to just:

`src/pages/api/auth/[...nextAuth].ts`:
```
export default NextAuth({
  ...
  callbacks: {
    session: async ({ session, token }) => {
      if (session?.user) {
        session.user.id = token.sub;
      }
      return session;
    },
  },
  session: {
    strategy: 'jwt',
  },
  ...
});
```

`types.d.ts` (file in project root):

```
import type { DefaultUser } from 'next-auth';

declare module 'next-auth' {
  interface Session {
    user?: DefaultUser & {
      id: string;
    };
  }
}
```

`token.sub` already exists and it's exactly the `user.id` I needed: https://github.com/nextauthjs/next-auth/blob/d1d2d977fefc1200b1ddb4bd665b405e5ae64d1e/src/core/routes/callback.ts#L127

**Reply by KirschX** [reply]
Thank you so much, You saved my time.

**Reply by reevebarreto** [reply]
Thanks a lot! I can breathe a lot better now

**Reply by dawkaka** [reply]
Put the `types.d.ts` file in the pages directory to advantage of Next.js's built-in TypeScript configuration and avoid extra configuration steps.

**Reply by mengxi-ream** [reply]
For people who need to use jwt, type declare should be module `next-auth/jwt` not `next-auth/jwt/type`
```ts
declare module "next-auth/jwt" {
  interface JWT {
    uid: string;
  }
}
```

I make the uid as `provider-id`, e.g. `google-126786142869245`, `github-12512345135`. I did it this way because I don't know if two OAuth providers has the same user id or not. And also I treat users with the same email but different providers as different users

so my code is like
```ts
    async session({ session, token }) {
      if (session?.user) {
        session.user.uid = token.uid;
      }
      return session;
    },
    async jwt({ token, account, user, profile }) {
      if (user) {
        token.uid = `${account?.provider}${account === undefined ? "" : "-"}${
          user.id
        }`;
      }
      return token;
    },
  },
  session: {
    strategy: "jwt",
  },
  ```
  and in `type.d.ts`
  ```
  import type { DefaultUser } from "next-auth";

declare module "next-auth/jwt" {
  interface JWT {
    uid: string;
  }
}

declare module "next-auth" {
  interface Session {
    user?: DefaultUser & {
      uid: string;
    };
  }
}
```

**Reply by mengxi-ream** [reply]
note if you want to use custom id in the session, don't use getServerSession in Layout. https://github.com/nextauthjs/next-auth/discussions/536#discussioncomment-7039765

**Comment by bperunx** [comment]
You are my hero, @ruohola. THX a lot!!! Your solution was working for me.
Also thanks to all other guys from that thread to figure out the solution.

**Comment by markmendozadev** [comment]
Hi

i'm trying out CredentialsProvider and when i'm trying to return the id it doesn't show up.

the only things show up is (name,email,image) is it possible to get the _id from mongo db?


```
import NextAuth from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import { verifyPassword } from "../../../lib/auth";
import { connectDb } from "../../../lib/db";

export default NextAuth({
  providers: [
    CredentialsProvider({
      async authorize(credentials) {
        const { email, password } = credentials;
        const client = await connectDb();
        const db = client.db();

        const user = await db.collection("users").findOne({ email: email });
        if (!user) {
          client.close();
          throw new Error("Email already exists");
        }

        const isValid = await verifyPassword(password, user.password);
        if (!isValid) {
          client.close();
          throw new Error("Could Not Log you in");
        }
        return user;
      },
    }),
  ],
  callbacks: {
    session: async ({ session, token }) => {
      if (session?.user) {
        session.user.id = token.sub;
        console.log(token);
      }
      return session;
    },
  },

  session: {
    strategy: "jwt",
  },
});

```

**Reply by shadoath** [reply]
Did this work?

**Comment by nickretallack** [comment]
Why isn't this in the example app?  Auth is pretty useless if you don't remember the user id of the person who logged in, right?

Do you have a solution for non-JWT sessions?  Is that not something the database adapters (e.g. prisma) handle automatically?

**Reply by youngnishant** [reply]
I used JWT lib for both authentication and authorisation, in next js app

**Reply by youngnishant** [reply]
Not maintained sessions

**Reply by youngnishant** [reply]
Use cookies, bestest solution for next

**Comment by nickretallack** [comment]
Here's what I came up with for those who don't want to use JWT.

``` typescript
// pages/api/auth/[...nextauth].ts
import NextAuth from "next-auth"
import TwitterProvider from "next-auth/providers/twitter"
import prisma from "@/lib/prisma"
import { PrismaAdapter } from "@next-auth/prisma-adapter"

export default NextAuth({
  adapter: PrismaAdapter(prisma),
  providers: [
    TwitterProvider({
      clientId: process.env.TWITTER_ID,
      clientSecret: process.env.TWITTER_SECRET,
      version: "2.0",
    }),
  ],
  session: {strategy: "database"},
})

```
```typescript
// pages/api/test.ts
import prisma from "@/lib/prisma"
import { getCookie } from 'cookies-next';

export default async function handler(req, res) {
  const token = getCookie('next-auth.session-token', {req, res}) as string
  const session = await prisma.session.findUnique({where: {sessionToken: token}})
  res.status(200).json(session)
}
```
```typescript
// lib/prisma.ts
import { PrismaClient } from "@prisma/client"
const prisma = new PrismaClient()
export default prisma
```

This uses the `cookies-next` library.  `next-auth` already uses the `cookie` library, so if you'd prefer to have less dependencies you could do this:
```typescript
// pages/api/test.ts
import prisma from "@/lib/prisma"
import cookie from 'cookie';

export default async function handler(req, res) {
  const cookies = cookie.parse(req.headers.cookie || '')
  const token = cookies['next-auth.session-token']
  const session = await prisma.session.findUnique({where: {sessionToken: token}})
  res.status(200).json(session)
}
```

**Reply by youngnishant** [reply]
Awesome ⭐

**Reply by beaumontyun** [reply]
Exactly what I am looking for. Thank you.

**Reply by michael-mcgreal** [reply]
Superb! Many thanks @nickretallack

**Reply by michael-mcgreal** [reply]
Note: In HTTPS production environments, next-auth will use a [secure cookie](https://next-auth.js.org/configuration/options#usesecurecookies) by default, which has a slightly different cookie key:

  ```
 process.env.NODE_ENV == 'development'
      ? cookies['next-auth.session-token']
      : cookies['__Secure-next-auth.session-token']
```

**Reply by felipeotarola** [reply]
best! 
this worked for me also.

    const session = await prisma.user.findFirst();
    console.log(session?.id);

**Comment by leephan2k1** [comment]
This code worked for me, now I'm on next-auth version ^4.6.1.  Then the object user is completely undefined in the jwt callback and session. I logged the check and the user id exists again inside the token's sub. Totally fine with me!

```
    callbacks: {
        async session({ session, token }) {
            session = {
                ...session,
                user: {
                    ...session.user,
                    id: token?.sub,
                },
            };
            return session;
        },
    },
```

**Reply by arvl130** [reply]
**UPDATE**: I think I had a misunderstanding while debugging this. The marked [solution](https://github.com/nextauthjs/next-auth/discussions/536#discussioncomment-1863534) in this thread—which is a variation of the solution above—actually solves it for me.

This is what works for me as well. Though, I'd just like to add that the user object, does show up on the `jwt` callback. Strangely, the callback just happens to run twice, and on the second run, the user object is no longer there, only the generated token is left. You can observe it here:

`[...nextauth].ts`
```
import NextAuth, { type NextAuthOptions } from "next-auth"
import CredentialsProvider from "next-auth/providers/credentials"

export const authOptions: NextAuthOptions = {
  callbacks: {
    jwt(jwtProps) {
      console.log("jwtProps", jwtProps)
      return jwtProps.token
    },
    session(sessionProps) {
      console.log("sessionProps", sessionProps)
      return sessionProps.session
    },
  },
  session: {
    strategy: "jwt",
  },
  providers: [
    CredentialsProvider({
      name: "Credentials",
      credentials: {
        username: { label: "Username", type: "text" },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials) {
        if (
          credentials?.username === "admin" &&
          credentials?.password === "admin"
        )
          return {
            id: "1234",
            name: "J Smith",
            email: "jsmith@example.com",
            some: "a",
            custom: "b",
            properties: "c",
          }

        return null
      },
    }),
  ],
}

export default NextAuth(authOptions)
```

Output after running `signIn()`:
```
jwtProps {
  token: {
    name: 'J Smith',
    email: 'jsmith@example.com',
    picture: undefined,
    sub: '1234'
  },
  user: {
    id: '1234',
    name: 'J Smith',
    email: 'jsmith@example.com',
    some: 'a',
    custom: 'b',
    properties: 'c'
  },
  account: {
    providerAccountId: '1234',
    type: 'credentials',
    provider: 'credentials'
  },
  isNewUser: false
}
jwtProps {
  token: {
    name: 'J Smith',
    email: 'jsmith@example.com',
    sub: '1234',
    iat: 1672392394,
    exp: 1674984394,
    jti: '89f57295-0db0-44c4-9de7-0f538d162ac1'
  }
}
sessionProps {
  session: {
    user: { name: 'J Smith', email: 'jsmith@example.com', image: undefined },
    expires: '2023-01-29T09:26:34.834Z'
  },
  token: {
    name: 'J Smith',
    email: 'jsmith@example.com',
    sub: '1234',
    iat: 1672392394,
    exp: 1674984394,
    jti: '89f57295-0db0-44c4-9de7-0f538d162ac1'
  }
}
```

I'm not sure how we can pass the user object along with the generated token, so I've taken a similar approach.

**Comment by kristianeboe** [comment]
@balazsorban44 @nickretallack @leephan2k1 

I wanted not just the userId, but also the additional fields I had added to my user model and ended up with, what seems to me at least, an even simpler solution:

```
export default NextAuth({
  adapter: PrismaAdapter(prisma),
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID!,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
    }),
  ],
  secret: process.env.NEXTAUTH_SECRET!,
  callbacks: {
    session: async ({ session, user }) => {
      return {
        ...session,
        user: user,
      };
    },
  },
});
```

**Reply by beaumontyun** [reply]
@kristianeboe Your solution was extraordinary at how simple and elegant it was. Seeing the "id" field at the console right out of useSession on a Monday morning made the week for me. Thank you.

**Reply by cobbvanth** [reply]
This is the proper solution, especially if you're not using JWTs

**Reply by cobbvanth** [reply]
any idea how to fix the typeerror? the NextAuthOptions doesn't like returning a user there

**Reply by AndrewRayCode** [reply]
If you use this solution, keep in mind this will return the entire user object to your client. If you add additional fields to the user model in your database that are sensitive, like a hashed password, it will be available in to any client call with `useSession()`.

It might be safer to pick the specific keys you want. Here's the solution I have now, including the proper typescript bindings (I use this solution to get the user id in the session)

```
import NextAuth, { User, Session } from 'next-auth';

....

  callbacks: {
    session: async ({ session, user }: { session: Session; user: User }) => ({
      ...session,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        image: user.image
      },
    }),
  },
```

**Reply by eyadaldaoud** [reply]
> @balazsorban44 @nickretallack @leephan2k1
> 
> I wanted not just the userId, but also the additional fields I had added to my user model and ended up with, what seems to me at least, an even simpler solution:
> 
> ```
> export default NextAuth({
>   adapter: PrismaAdapter(prisma),
>   providers: [
>     GoogleProvider({
>       clientId: process.env.GOOGLE_CLIENT_ID!,
>       clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
>     }),
>   ],
>   secret: process.env.NEXTAUTH_SECRET!,
>   callbacks: {
>     session: async ({ session, user }) => {
>       return {
>         ...session,
>         user: user,
>       };
>     },
>   },
> });
> ```

Thank you so much , this worked for me and its pretty simple

**Comment by mattmcegg** [comment]
If anyone is wondering how to do this on Next.js 13 using NextAuth 4.22.1 using TypeScript, where I am returning data from the database about the user, including ID and roles, here ya go:

First, you have to make some custom types inside your types folder (you may need to edit your tsconfig.ts if you're not defining typeRoots by adding): 

`"typeRoots": ["./node_modules/@types", "./src/app/types"],"`

then, create a file called /next-auth.d.ts mines inside '/src/app/types' (or whatever folder your types are in):

```
import { Session as NextAuthSession, User as NextAuthUser } from "next-auth";

declare module "next-auth" {
  export interface User extends NextAuthUser {
    id: string;
    roles: string[];
  }

  export interface Session extends NextAuthSession {
    user: User;
  }
}

```

Then you need this inside [...nextauth].js:
```
interface ExtendedUser extends NextAuthUser {
  id: string;
  roles: string[];
}
```

then, fire away on the session callback:
```
async session({ session, token }) {
      const user = await prisma.user.findUnique({
        where: {
          id: token.sub,
        },
        select: {
          id: true,
          roles: true, // Include roles in the user query
        },
      });

      session.user = {
        ...session.user,
        id: token.sub || (user?.id as any),
        roles: user?.roles as any, // Add roles to the session.user object
      } as ExtendedUser; // Cast the session.user object to ExtendedUser

      return {
        ...session,
      };
    },
```

**Reply by simonasbuj** [reply]
wow, finally an answer that works, thank you! For me it worked in even simpler way.
To have full user data (username, image etc.) in useSession() I did this:

In root folder create a file 'next-auth.d.ts with this content:
( TO BE HONEST, I dont even know if this one is needed, works without it too)
```
import NextAuth from 'next-auth';

declare module 'next-auth' {
  interface Session {
    user: {
      id: string
    } & DefaultSession['user']
  }
}
```

in [...nextauth]/route.ts add callbacks section:
```
callbacks : {
        session: async ({ session, token }) => {
            const fullData = await prisma.user.findUnique({
                where: {
                    id: token.sub
                }
            })
            if (session?.user) {
              session.user.id = token.sub
              session.user.fullData = fullData
            }

            return session;
          },
          jwt: async ({ user, token }) => {
            if (user) {
              token.sub = user.id;
            }
            return token;
          },
    },
```

Now I can get any user data in components like this:
```
const session = useSession()
...
 <p>logged in user username: <span className="font-bold">{session.data?.user?.fullData.username}</span></p>
```

**Comment by karmalia** [comment]
![screen-shot](https://github.com/nextauthjs/next-auth/assets/100145738/c27d287a-b411-4dca-a02f-ee1dfc229ea2)

Fix: Convert your server side session to client and don't pass session to SessionProvider. Somehow your custom session is overwriting by this session object. If there is a good solution I also would like to know.

PS: I tried types trick. It only solved type checking. 

```
export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="tr">
      <body>
        <SessionProvider>
          <div className="flex">
            <Sidebar />
            <main className="relative max-h-[100vh] flex-1 overflow-y-auto">
              <Header />
              <div className="content">{children}</div>
            </main>
          </div>
          <ToastContainer />
        </SessionProvider>
      </body>
    </html>
  );
}

```

**Reply by mengxi-ream** [reply]
You are write, we should not use getServerSession in Layout and pass it to Provider, otherwise it will overwrite our custom id in session.
```tsx
"use client";
import { SessionProvider } from "next-auth/react";

const SessionProviderWrap = ({ children }: { children: React.ReactNode }) => {
  return <SessionProvider>{children}</SessionProvider>;
};

export default SessionProviderWrap;
```
```tsx
import "./globals.css";
import type { Metadata } from "next";
import { Inter } from "next/font/google";
import Navbar from "@/components/nav/Navbar";
import SessionProviderWrap from "./SessionProviderWrap";

const inter = Inter({ subsets: ["latin"] });

export const metadata: Metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body className={`${inter.className} min-h-screen flex flex-col`}>
        <SessionProviderWrap>
          <Navbar />
          <main className="-mt-32 flex-grow flex">
            <div className="mx-auto max-w-7xl px-4 pb-12 sm:px-6 lg:px-8 flex-grow">
              <div className="rounded-lg bg-white px-5 py-6 shadow sm:px-6 min-h-full">
                {children}
              </div>
            </div>
          </main>
        </SessionProviderWrap>
      </body>
    </html>
  );
}
```

**Comment by vyperrrr** [comment]
I have the same problem (sort of). I am able to get userId using useSession() hook, but getServerSession() doesn't include userId.
![image](https://github.com/nextauthjs/next-auth/assets/78800814/8190146f-2ea9-427f-8d77-898df17b51e2)
